### DISCUSSION TITLE
Making PUDL into a Data Warehouse

### DISCUSSION THREAD

**Discussion by bendnorman** [catalyst-cooperative/pudl#1838] [Ideas]
PUDL has two main parts, the ETL and what we call the output/analysis layer. The [ETL](https://catalystcoop-pudl.readthedocs.io/en/dev/autoapi/pudl/etl/index.html) extracts poorly structured data from excel, dbf, and xbrl files and transforms the data into well-structured relational tables. The outputs tables are a collection of tabular outputs that contain the most useful core information from the PUDL data packages, including additional keys and human-readable names for the objects. You can read more about the [output](https://catalystcoop-pudl.readthedocs.io/en/dev/autoapi/pudl/output/index.html) and [analysis](https://catalystcoop-pudl.readthedocs.io/en/dev/autoapi/pudl/analysis/index.html) layers in the docs. Currently, the data these layers produce is only accessible via the pudl python package, but we are hoping to add many of the tables to a sqlite database we can share more easily. See issue #[1178](https://github.com/catalyst-cooperative/pudl/issues/1178).

We are [applying dagster abstractions](https://github.com/catalyst-cooperative/pudl/issues/1570) to existing code to leverage dagster CD, parallel processing of tasks, and documentation of our ETL and data assets.   Issue [#1487](https://github.com/catalyst-cooperative/pudl/issues/1487) contains our reasoning for using dagster. While we are adding dagster to our codebase, we also want to be thinking about how to improve our overall design. Issue [1401](https://github.com/catalyst-cooperative/pudl/issues/1401) contains helpful background on our plethora of design pain points.

**Main issues:**
- We don’t have a uniform recipe for adding new datasets to our ETL. Each dataset is extracted and transformed in different ways. See issues [1401](https://github.com/catalyst-cooperative/pudl/issues/1401).
- We don’t have a standard structure or criteria for adding new analyses. [This discussion](https://docs.google.com/document/d/1tj-aVwGSh2KhExRFV9AtewRAbENtePHp3iTvykXVGUE/edit#heading=h.5bpck09ibo9b) contains some background.
- The data pipeline is not well organized in terms of what kinds of processing happen when. See issues [1401](https://github.com/catalyst-cooperative/pudl/issues/1401).

**Comment by zaneselvans** [comment]
Hey Noon @silky @7imon7ays and Brent aka @turbo3136 

Rehashing #1401 a bit here, we have several types of tasks / data assets that should probably get slotted into the ETL process / data warehousing process somewhere, but I'm not really sure where, or what the best way to arrange the dependencies between them will be, or what best practices around organizing the process is.

Some of the big ones:

### Raw-ish extracted data
* Currently we don't put the raw extracted data into the database, and it's only ephemerally available in a lot of cases.
* On exception to this it he FERC Form 1, which we translate from an old VisualFoxPro DB into SQLite first.
* Raw Excel and CSV inputs are available, but we don't store the early versions of them as database tables.
* The various partitions of these datasets typically can't be concatenated together without doing some data cleaning first (renaming columns, ensuring columns have a single compatible type)
* Tools like Airbyte seem to want to store every single partition as its own initial standalone table. That would be hundreds of tables and seems like kind of a mess.
* What's the first state of the data that should get written to a DB? Right now the first step that would be easy is when we have all the partitions concatenated together with uniform column names, but the data types would still be unruly.
* The raw FERC 1 DB (and all the new FERC XBRL data, which we're also able to turn into standalone SQLite DBs) could also easily be loaded into data warehouse tables.

### Structural data repair or integration processes
We have a couple of ML-lite processes that we use to knit different datasets together. Right now this work happens in a bunch of different places, which is kind of a mess.
* Identifying related records from multiple years of FERC 1 plant data, by calculating the cosine similarity of vectorized records and choosing unique sets of closely aligned records. Currently this happens very early in the FERC 1 transformation process and probably could / should be deferred until later. It also uses ~24GB of memory at the moment, and scales quadratically with the number of FERC steam plant records 😬.
* Identifying aggregations of EIA plant records that are mostly likely to correspond to the plant data that has been reported in FERC Form 1. Currently this uses the Python Record Linkage Toolkit and a logistic regression model. This currently happens in the "analysis" layer that runs on top of the database (and also takes like 45 minutes to run from scratch and uses too much memory to run in CI, which @katie-lamb is trying to address)
* Linking depreciation data collected from state PUC filings and FERC 1 to the appropriate EIA plant records. This is currently a fuzzy string-matching process that like the record linkage process above runs on top of the database.
* EIA data reconciliation & normalization (aka "harvesting") which deduplicates multiply reported values from all across the EIA datasets, and generates "golden records" that are stored in entity tables representing plants, utilities, generators, boilers, balancing authorities, emissions units, cooling systems etc. -- the natural building blocks of the electricity system that the data represents.

### Analysis
The "analysis" tables are typically the results of more complex calculations, that I think we'll probably want to keep using Python for. They depend on lots of other data in the database, and will end up creating a small number of new additional columns that should usually have the same primary keys as existing entities (e.g. monthly per-generator records). This includes stuff like:
* Calculating the marginal cost of electricity (MCOE) on a per-unit, per-month basis, which involves some messy heuristic allocation of fuel consumption and net generation across boilers and generators for which we only have aggregated data.
* Estimating hourly state-level (or someday county-level) historical electricity demand
* Identifying outlying values in collections of correlated time series, and replacing them with reasonable values
* Estimating the hourly carbon intensity of electricity generation within a certain geography or jurisdiction based on which units are operating at what output.

### Denormalized output tables / database views
* What we've called the "output" tables seem like they can be re-implemented as materialized views within the database / warehouse, implemented entirely using SQL (though we may need to bone up on SQL).
* Right now, because we've historically developed most of our calculations interactively using these output tables for convenience, most of our analyses depend on them.
* This seems like it might be the opposite of what we want in production though, since we'll want many derived values from more complex calculations to be available in the denormalized views for convenient use.
* Or maybe there are several different layers of denormalized tables? Some that refer to the original data, and others that are built from the derived values later on?

**Reply by silky** [reply]
Can I ask a wildcard question - have you given much thought/consideration to GraphQL backends? I've spent a fair bit of time playing around with [Hasura](https://hasura.io/), and I really _really_ like it. In particular I think they put a lot of effort into the ability to "link" things; so I think it could be well-suited for this space; i.e. if you go down the path, that I think is pretty reasonable, of making tables for everything (so it's saved); but then not getting too overwhelmed about how to keep them all linked and connected.

I haven't used it for this exact purpose; but it could be fun/fruitful to investigate/experiment.

**Reply by zaneselvans** [reply]
We haven't looked at using GraphQL at all in any context, but the founder of Dagster was one of the creators of GraphQL at Facebook and I think there's a GraphQL API of some kind built into Dagster for tracking the relationships between different data assets and transformations. So I think by virtue of defining the relationships through Dagster's abstractions we'll end up with a GraphQL representation of the DAG.

Is Hasura analogous to something like FastAPI? For defining an API for external access to an existing database? We aren't (yet) trying to do that.  We're still trying to just create the database itself, and step through the various kinds of transformations required. We're also trying to avoid getting tangled up in a bunch of proprietary services if we can.

**Reply by silky** [reply]
I've never used FastAPI; Hasura is an (open-source; free if self-hosted, as I've used it before) GraphQL "server"; so what you do is connect it up to existing databases ( https://hasura.io/docs/latest/databases/index/ ) and then you can get unified querying interfaces. You _can_ set it up for public consumption, but it can just as well be used for your own internal querying.

One of the reasons I'm suggesting thinking about it, is because it could mean you can kind of defer finding a "perfect" essential database structure, and just connect things up a bit more dynamically. 

For example, what I'm thinking is that because with something like this you can support queries across multiple databases simultaneously, easily, the "easiest" option is just to make specific databases for each of the specific datasets, and then have this kind of "unifying" interface; that can then be versioned/deprecated/etc as time goes on.

But maybe I'm totally on the wrong mindset here ... could well be.

**Reply by zaneselvans** [reply]
Our model so far as been *not* to try and host a live remote DB that users access or build on top of. Instead we archive ([on Zenodo](https://zenodo.org/communities/catalyst-cooperative/)) and distribute (via [an Intake data catalog](https://github.com/catalyst-cooperative/pudl-catalog)) complete, versioned SQLite DBs and/or Parquet datasets, so that we can avoid the cost and responsibility of maintaining that live infrastructure, and so people don't have to work with the data over a network connection. At the moment we're only talking about a few GB of data, which is pretty convenient for a user to download once in bulk (whenever a new version is released). We've been using [Simon Willison's Datasette](https://datasette.io/) to provide a more point-and-click interface to the DBs (since publishing them to Cloud Run is so easy).  See https://data.catalyst.coop

Totally agree that if/when we are presenting a public API that we want people to build on top of, it would be great to decouple the internal structure from the external interface!

But for now, we're mainly talking about creating a "data warehouse" in the context of the data processing pipeline, for use in CI/CD, so that it has a place to persist useful intermediary data assets which, at the end of the process, we can export to the file-based assets for distribution. Of course once we've got this live, continuously updating DB, I'm sure we'll start thinking about stuff we can do with it (especially some visualizations!)

This model also seems to be getting some traction with the development of [DuckDB](https://duckdb.org/), which is kind of a columnar / OLAP DB that mirrors a lot of the patterns established by SQLite. There was interesting discussion [on the Data Engineering Podcast a few months ago](https://www.dataengineeringpodcast.com/duckdb-in-process-olap-database-episode-270/).

**Reply by TrentonBush** [reply]
>Tools like Airbyte seem to want to store every single partition as its own initial standalone table. That would be hundreds of tables and seems like kind of a mess.

As long as we put them in their own schema/namespace, I think even a bazillion tables would be easy to ignore

**Comment by zaneselvans** [comment]
Another thing I'm really unclear on with column-oriented Data Warehouses is how people deal with relationships between tables. It seems like data normalization and foreign key constraints kind of get tossed out the window, but these seem really important to me for knowing / requiring that the data really has the structure we think it has, that there's only one source of truth, and allowing us to find errors / outliers to correct. How do folks typically represent these constraints?

Do you typically have a nice normalized original database, which is then used to derive all the other denormalized tables downstream?

**Reply by silky** [reply]
@zaneselvans , probably I'm missing something, but I don't think that column-oriented typically means no relationships? What brought about that thought? 

(I'm thinking mostly of [parquet](https://arrow.apache.org/docs/python/parquet.html#obtaining-pyarrow-with-parquet-support), which in the past I've been a consumer of, but I don't recall these issues?)

**Reply by zaneselvans** [reply]
We use Parquet for the billion-row EPA CEMS data, but we don't have any foreign key constraints between that table and the database. And as far as I can tell they don't support FK constraints between different tables stored in Parquet. How would one specify and enforce them? BigQuery doesn't support primary / foreign key relationships. And I've just seen this referenced in various places. It seems really weird to me.

**Reply by silky** [reply]
> How would one specify and enforce them?

Yeah; good question, I guess you're right. From glancing around, it does seem that people just do the joining "manually", as it were. I guess it probably makes sense that the constraints aren't enforced (the price you pay for them being separable?) 😳

**Reply by zaneselvans** [reply]
I think the main benefit that you get is the big speedup in analytical workflows, where you're primarily doing type-homogeneous operations on columns, which can be very efficiently vectorized in memory, and often distributed across multiple cores or computation nodes.

But so far in our experience, direct operations on SQLite (despite not being columnar) has been *plenty* fast for interactive computing, and the main bottlenecks in performance have been the fact that we're often doing SQL-ish stuff in pandas (maybe not as efficiently as we could), and repeating the same computations many times.  The goal with the data warehouse / Dagster refactor is to get all of those useful, frequently recomputed assets into the DB itself, so that all you have to do is `SELECT * FROM ...` or `pd.read_sql()` etc. from whatever computational platform you like, without needing to deal with all of our python code and dependencies.

The only exception so far as been the EPA's Continuous Emissions Monitory System (CEMS) hourly emissions data, which is about a billion rows, and way too big for in-memory analysis on a laptop.  There will be a few other long skinny tables like this that we've decided to store in Parquet. We know that the FERC Electric Quarterly Report will also fall into this category, and a lot of the grid operator data would too, if we decided to pull it in.

We'll need to somehow do constraint checking in the ETL process though, so that we know the data in these long tables is consistent with the bulk of the relational data stored in the SQL DB. And in both cases (but especially the EQR) there's a bunch of entity matching that is going to have to be done on the long tables to make them really useful in conjunction with the other data. And that's a major issue that's pushed us off of our laptops and into cloud infrastructure / orchestration frameworks.

**Reply by bendnorman** [reply]
I think you typically have a nice normalized original database, which is then used to derive all the other denormalized tables downstream.

Typically, data warehouses' source data is transactional data that is normalized and specifies relationships. The transactional data is processed through an ETL and loaded into a data warehouse with a more denormalized schema like a star schema that is conducive to analysis.

This general workflow doesn't really fit for PUDL because our source data is so messy and lacks structure. Our current ETL gets the raw data to a point where it looks like transactional data: normalized, constraints, relationships... etc. The PUDL output tables feel like the denormalized tables of a data warehouse that we can do analysis with. 

**Do y'all think this is an accurate description of how PUDL's architecture maps to standard data warehouse architectures?**

It feels like the raw data's messiness is the origin of our architecture confusion. Data goes through multiple extract, transform and load steps until it's ready for analysis. For example, the Ferc Form 1 data is extract from dbf files and loaded into the ferc1.db database. Then it is extracted again from ferc1.db and transformed and loaded into our pudl.db. The cleaned ferc tables are extracted again from pudl.db and transformed into denormalized tables analysts can use. 

**How would y'all use common data engineering patterns to redesign the FERC Form 1 pipeline? Does the messiness of the raw data create a unique situation where common patterns don't apply?**

**Reply by turbo3136** [reply]
Hey all! Sorry I'm late to the party here, but I'll try to follow up on Ben's questions here.

> **Do y'all think this is an accurate description of how PUDL's architecture maps to standard data warehouse architectures?**

Summarizing your comment about standard architectures, I think you're saying that most ETLs/ELTs start with data that looks transactional (a thing happens at a certain time and info about that thing is captured in a row) and ends up more denormalized (info about the thing is also connected to other relevant info to reduce the complexity of analytical queries). I think that's pretty much spot on. Only thing I'd add is that there's often a little more cleaning at the beginning and in the warehouses I'm familiar with we've also built out a lot more layers of analytical tables that are even more denormalized than a star schema that you can plug straight into a dashboard or regression model, which may or may not be applicable to you.

> **How would y'all use common data engineering patterns to redesign the FERC Form 1 pipeline? Does the messiness of the raw data create a unique situation where common patterns don't apply?**

Your data is definitely not so messy that you're in a unique situation! I'm pretty confident we can make common patterns apply here. The way I've broken this problem down in the past is to clearly document the pieces of information that I want to clean up and whether I want to make them look more like transactional data, reference data, or leave them as is. In other words, I look at the really messy data I'm starting with and ask myself "In a perfect world, how would I want this data to look when it's given to me?" Then I do that processing. Now we're at the normal spot of having transactional data and reference data that we want to combine into data warehouse tables that are more useful to analysts.
Given that I'm way more comfortable in SQL than pandas or dask, I generally follow the ELT pattern instead of ETL, meaning the data gets sent to the data warehouse and then all the transformations are done in SQL. Because of that, I usually have 5 separate steps in my data processing, which I do all in SQL:
1. protective views: this is usually helpful in my use-cases but might not be for you. I often put a "protective" layer in between the raw data I consume and the processing I do on top of it. I create a bunch of `SELECT * ...` views of the raw data. This is extremely useful in the case where the raw data changes slightly, like a column name is updated or a string's value changes. I can just make the fix in this protective layer, then the rest of my ELT works as normal.
2. pre_processed tables: clean up the messiness of the data and get it into flat tables that look more like traditional relational data. This is where all of your really messy transformations happen.
3. elemental tables: only pass through the pieces of information I care about (the transactional and reference data) that I cleaned up. These tables are still normalized and ensure a single source of truth for every piece of information I care about. Maintaining a SSoT is really important for efficiency because you only have to update logic once, but also because multiple sources of truth cause confusion for users and lead to incorrect analyses. That's another reason I find this step useful because the pre-processing step often involves some intermediary data that I don't want to expose to users.
4. analytics (catalyst) data mart: here's where I do a lot of denormalization and combine data to make it more useful for analysts. I generally make these tables very wide and break the SSoT rule because these transformations do not add any logic to the data, they just join and re-arrange data. So I will shove the same piece of information into 10 different tables if the analysts find it useful. I also called this the "catalyst" data mart because I usually make this as generally useful for all analysts as I can. It would probably be consumed by internal stakeholders. Data in here can be actual tables, views, or materialized views depending on the use case.
5. project-specific data marts: I'm not sure if this will be useful to you, but after the analytics data mart I also usually create data marts specific to certain teams or projects. For example, I might have separate data marts for my Marketing team and my Engineering team that have different data specific to each team. I keep these separate so that I can only expose certain data to certain teams. Too much data gets confusing.

To be perfectly clear, I did not come up with this myself. I stole most of these ideas from [this video from the founder of Periscope Data](https://youtu.be/D5hpjlYHEGw?t=386).

The summary of the FERC Form 1 pipeline you gave makes it sound like you're already doing something a somewhat similar, but not exactly the same. If I understood right, your process looks like:
1. raw data
2. ferc1.db: only ferc data, heavily processed and cleaned from its messy state into something more usable, probably maintains the SSoT for all ferc-related data points
3. pudl.db: combines all data sources together into one db, still maintaining SSoT
4. analytics data: denormalized for analytics

My guess is it's extra complicated because (1) the other processing from other data sources doesn't happen in the exact same format, (2) doing all of these transformations in separate python functions is messy and harder to follow than in SQL (at least for me!), and (3) you have to move data in and out of databases instead of having all the transformations happen within the same database. If I'm right about the complications you're experiencing, I think there are a few ways to simplify all this. It probably depends on exactly _why_ you want to simplify it. Is it for efficiency of updates and reliability, is it to follow modern best-practices and career development, is it breaking a lot and takes a week to fix, or something else? I'm asking about the why because it's entirely possible that lots of the solutions I might suggest (like working more in SQL, keeping data within the same database, etc) will make your lives even harder in the long run because I'm solving for a different problem than the one you're experiencing. So I guess the simplest version of my question is: why do you want to apply common data engineering patterns to redesign the FERC Form 1 pipeline? I'm guessing the answer is these 3 problems you mentioned above:
- We don’t have a uniform recipe for adding new datasets to our ETL. Each dataset is extracted and transformed in different ways. See issues https://github.com/catalyst-cooperative/pudl/issues/1401.
- We don’t have a standard structure or criteria for adding new analyses. [This discussion](https://docs.google.com/document/d/1tj-aVwGSh2KhExRFV9AtewRAbENtePHp3iTvykXVGUE/edit#heading=h.5bpck09ibo9b) contains some background.
- The data pipeline is not well organized in terms of what kinds of processing happen when. See issues https://github.com/catalyst-cooperative/pudl/issues/1401.

OK, that was a lot of text! Let me know if anything is unclear or if I missed the point of your questions. To highlight my only real question in all this text: **did I get the 3 main reasons you want to apply common data engineering patterns to your transformations right, from the paragraph above?** If this wall of text isn't clear enough or you want to chat live about any of this, let me know. I think that might be helpful given the complexity of what we're trying to do here.


Side-note for @zaneselvans and your question about primary/foreign keys. You're right that these constraints are almost always ignored in modern data warehouses. That's primarily because (1) they're not really needed in column stores and don't change efficiency much and (2) in the highly-denormalized world of data marts (denormalized beyond even a star schema) you don't really need to do extremely complex joins and you've already renamed your columns to have obvious join conditions, plus you included a lot of uniqueness tests in the elemental step that mimic the constraints from primary/foreign keys. I think that last part answers your concern. Basically, the tradeoff is to replace the primary/foreign key constraints and the automated testing associated with them for uniqueness and other tests in the "elemental" processing step. I personally don't find this to cause problems for analysts or general consumers of the data, but that could just be me and the people I work with.

**Reply by bendnorman** [reply]
Thank you for all of the info! Yes, I think those are three of the main reasons we are looking to adopt modern data patterns. We are looking for common patterns because each data pipeline and piece of analysis feels bespoke. It would be nice to reference tested patterns when adding new data and analyses, so we feel confident in our ability to maintain the data and code. @zaneselvans can probably elaborate on this.

I think PUDL's existing design shares some parallels with your process. Our ETL step and your pre_processed clean up the messiness of the data and get it into a normalized database. Our output layer and your analysis data marts denormalize the cleaned data for analysts to use. 

Obviously, the implementation of this pattern will look quite different in python and SQL. We are mostly familiar with python, and we don’t know if certain cleaning steps like ML entity resolution steps can be done in SQL. We are hoping to convert our output layer into database views. 

I have a couple of additional questions:
- Can you talk a little more about elemental tables? Are they just the output of the pre_processed tables? 
- Sometimes analysis needs to be integrated into the data cleaning. For example, someone might develop an imputation model for a field. What do you do when analysis performed on a data mart table should be integrated into the pre_processing step?

**Reply by turbo3136** [reply]
> - Can you talk a little more about elemental tables? Are they just the output of the pre_processed tables?

Yes, they are the output of the pre_processed tables. The only reason I've separated this in previous warehouses is for user-side (or analyst-side) simplicity. There may be fields or tables you used in pre_processing but don't want to include in any user-facing data because they're not useful or cause confusion. In my case, this is also helpful because I have product analysts (who are not familiar with the rest of the data pipeline) that consume elemental tables to build data marts for their own analyses. If this isn't an immediate concern for you I'd probably ignore it for now.

> - Sometimes analysis needs to be integrated into the data cleaning. For example, someone might develop an imputation model for a field. What do you do when analysis performed on a data mart table should be integrated into the pre_processing step?

Yes, this definitely happens to me frequently! I'd say there are three different approaches I take here depending on the specific problem:
1. If the data to be integrated is dependent on a one-time analysis, I'll often just create a table with the results of the analysis (and document the analysis somewhere). Basically, if it makes sense to hard-code the results I'll do that. 
2. If the analysis can be done on the data available in the pre_processed step, I just pull the imputations into the pre_processed step as though it were part of the normal transformations.
3. If the analysis can't be done on the data available in the pre_processed step, but I need it in other calculations and want to pull it into the pre_processed step (usually to preserve a SSoT), I usually bring the whole set of transformations that created the data mart table into pre_processed. It's possible this is more difficult for you because you maintain separate databases here and your imputed values might come from more than one of those databases. Even then, I think there are solutions we could explore.

I'm guessing your imputation situation falls into (3) above. Either way, I'd be happy to take a look at that specific code if you'd like me to comment on it. If the imputation example is a real one, do you mind sharing the link to the code? I think that would be a great example for us to use here since I've only been talking in abstract terms so far. It would also help me more deeply understand and think through your specific problems.

**Reply by bendnorman** [reply]
Thanks! @cmgosnell @zaneselvans do y'all have good examples of analysis work that should be integrated into our main ETL?

**Reply by zaneselvans** [reply]
Finally getting back to this conversation! I had to go dive into Wikipedia to get more familiar with some of this vocab, which thankfully seems to map pretty directly to the problems we're trying to solve. I think they correspond roughly like:
* **[Fact table](https://en.wikipedia.org/wiki/Fact_table):** what we've internally often called "data tables". In EIA these 
* **[Dimension](https://en.wikipedia.org/wiki/Dimension_(data_warehouse)):** what we've internally often called "entities" (e.g. generators, utilities, plants, and balancing authorities) whose static or slowly varying attributes are stored in our "entity tables" (`plants_entity_eia`, `plants_eia860`, `generators_entity_eia`, `generators_eia860` etc.)
* **[Star schema](https://en.wikipedia.org/wiki/Star_schema):** In our output tables we do something approximately like this by bringing together the plant and utility tables (both their permanent and annually variable dimensions using the `pu_eia860()` function (which is mostly SQL-like stuff in pandas) before joining that information to the core data tables like `generation_fuel_eia923` or `generation_eia923` or `fuel_receipts_costs_eia923`
* **[Snowflake schema](https://en.wikipedia.org/wiki/Snowflake_schema):** This seems close to what we've been imagining producing as an intermediate normalized data product, before we start creating the denormalized views which would be more like Star schemas. But since we have a variety of different "fact tables" which are often referencing shared dimensions (entities) maybe it's more like **[Fact constellation or Galaxy schema](https://en.wikipedia.org/wiki/Fact_constellation)**.

### Clarifying the FERC 1 pipeline
@turbo3136 just to clarify your outline of the FERC1 pipeline a bit:
* The original FERC 1 DBF to SQLite operation is almost entirely a format conversion. It just unifies a bunch of annual Visual FoxPro DBs (a format nobody supports) into a single multi-year SQLite DB that we can access easily using modern tools.
* Similarly the new FERC 1 XBRL to SQLite operation translates the XBRL (a business reporting specific dialect of XML, which FERC has just started publishing via RSS) converts the nested, non-tabular XML data into another SQLite DB, so it can be queried normally.
* Either of these processes could dump into one big PostgreSQL or other DB instead. Storing and publishing this intermediate format is useful because there are 100+ tables, with 3000+ columns, covering almost 30 years reporting, and because the formats they use are kind of arcane, it's really hard for most people to access for analysis, but we can't clean it all up completely -- right now we only have like 7 of the tables pulled into PUDL, and at most we might get up to 20 or so by the end of the year. Having the other 80+ raw tables available in *some* modern format is valuable.
* Then we extract whole tables from the converted raw FERC 1 DBs and do the messy cleanup steps. Much of this is probably work that someone experienced with SQL could do directly in that language, but some steps aren't, and none of us are familiar with the tools that allow the parameterization and re-use of SQL procedures.
* One challenge we face is that often the complicated data cleaning we're doing is in the service of giving the data structure that it doesn't have to begin with. In the case of FERC 1, power plants have no IDs, and their names and what portion of a plant reports together changes from year to year. So we have an involved process that classifies records from many different years as likely to be associated with each other, and algorithmically assigns a "plant ID" that links those records. This process is coordinated by [`pudl.transform.ferc1._plants_steam_assign_plant_ids()`](https://github.com/catalyst-cooperative/pudl/blob/d294cb52b62b6adf18384b2bbb1be15c84667e6f/src/pudl/transform/ferc1.py#L1770). Right now it happens as part of the same group of transformations that takes the data from the FERC 1 SQLite DBs to the cleaned up PUDL DB, which I think corresponds to your "pre-processed" tables, which is probably too early. Should this kind of thing be done when creating the elemental tables? Or does it only belong in the analytical tables? If anyone wants to be able to do aggregations by plant or look at time series, they need this ID.
* Later on we do some messy entity matching to connect the FERC plant reporting to the EIA plant & generator reporting, and the only IDs we have connecting the two datasets are currently created by us and manually assigned. That's only possible because there's just a few hundred FERC utilities, and a few thousand plants that show up in both datasets, and it's a coarse grained connection that doesn't fully reflect the resolution of the EIA data, so there's another entity matching process that uses a logistic regression and some manually categorized training data to identify which individual generator+ownership slices available on the EIA side are most likely to correspond to whatever portion of a plant has been reported to FERC. This association is generally useful for building other analyses that involve both the kinds of data reported to FERC and EIA, so I imagine this would go into the "catalyst data mart" layer.
* Insofar as it's possible I think we want to avoid saving hard-coded analyses unless it's something that's actually manual and so can't be re-run (like the FERC to EIA plant and utility mapping, or the training data mentioned above).

### Other Analysis / ETL Examples
A lot of kind of work shows up in [the `pudl.analysis` subpackage](https://github.com/catalyst-cooperative/pudl/tree/d294cb52b62b6adf18384b2bbb1be15c84667e6f/src/pudl/analysis)
* We do some messy work to identify which boilers and generators are associated with each other within the EIA data. This adds a `unit_id` key to both the `boiler` and `generator` entities, which is necessary for calculating thermal efficiencies, and thus the cost effectiveness and emissions intensity of various sub-plant units. Right now this is happening in the production of the **elemental tables** but it could be that these boiler-generator-associations (BGAs) don't need to get built until the analytics data mart.
* One analysis that the BGAs go into is the estimating the marginal cost of electricity on a per-generator or per generation-unit basis. That calculation also results in monthly capacity factors and heat rates (which are both ratios) that are important for understanding the operations of the power plants. To me this feels like deriving a new "fact table" that's based on the originally reported data -- the values are numerical and clearly associated with entities that have other attributes in the database.  I've been imagining that this kind of calculation would happen after the elemental tables are loaded into the database, and would produce some new long tables with a primary key of (report month, generator ID) or (report month, generation unit ID), which would then be made available through a denormalized materialized view, alongside all of the other information that's associated with that entity and temporal granularity.
* Alongside that work we also have to allocate various quantities that are reported on different bases in different data sources.  E.g. only some generators report their generation directly. Others report aggregated plant-wide generation on the basis of prime mover and fuel, and we have to estimate the amount of generation (and fuel consumption) that should be associated with each collection of boilers and generators.
* We also do a bunch of reconciliation of internally inconsistent data values in the data normalization / deduplication process, as we produce the elemental tables right now.  E.g. if a plant reports its latitude & longitude sliiiiiightly differently from year to year, how to we choose a single canonical value to associate with the plant permanently? Or if there are different spellings of the plant name from year to year, or in different tables in the same year, which name do we want to retain for use in entity matching with other datasets?

### Foreign Key Constraints & Denormalization
* Reading about these various data warehouse schemas, the main point that came up for why one might want to use a more normalized structure was reduced storage space, which isn't very important these days given how cheap storage is.
* To me the most useful part of enforcing our many foreign key constraints has been the way it highlights incorrect assumptions about the structural relationships between different data and entities, or how it illuminates data quality issues (bad codes in columns that are supposed to use a controlled vocabulary, alphanumeric IDs that have leading zeroes and are supposed to be treated as strings, but get converted to integers, errors that have been made in manually or algorithmically assigned IDs, etc.)
* Are there common techniques for identifying this kind of data problem in a denormalized data processing environment? Beyond just requiring that the natural primary keys are present and unique within an elemental table?

**Reply by turbo3136** [reply]
@zaneselvans I just wanted to acknowledge receipt here. I was in the mountains away from internet this week, but I'll reply early next week. Apologies for being slow to respond!

**Reply by turbo3136** [reply]
To reply to a few specifics here:
- [pudl.transform.ferc1._plants_steam_assign_plant_ids()](https://github.com/catalyst-cooperative/pudl/blob/d294cb52b62b6adf18384b2bbb1be15c84667e6f/src/pudl/transform/ferc1.py#L1770) makes the most sense to me in the pre-processed step assuming you need the ID for other joins and transformations within your data processing (and not just for analytics afterwards). I have to create IDs in my data a lot too, and generally I do that in the pre-processed because I need them throughout the rest of my data transformations. If that's not the case for you, then it could work in your data marts. My overall goal with this structure is to keep things simple, decrease inter-dependencies, and never duplicate definitions. If you think about where this ID will be used and how it's created, will it result in a simpler DAG if you do it in pre-processed or later? My guess is it's simpler in the pre-processed.
- _Are there common techniques for identifying this kind of data problem in a denormalized data processing environment? Beyond just requiring that the natural primary keys are present and unique within an elemental table?_ Yes! I generally call these unit tests or integration tests. I use a tool called dbt (data build tool) that lets me include these tests in the documentation I create. [Here's dbt's example, see the lines with "tests"](https://docs.getdbt.com/reference/configs-and-properties#example). You can also do this in pytest or whatever tool you prefer. But... if you already have primary/foreign keys defined and those "tests" work well for you, I don't see a reason to change that.

To step back a bit, it sounds like the biggest problems you're running into is around the structure of your transformations and having a "recipe" for adding new ones. It's easy to argue that you can do this kind of organization yourself with directory and file structures, plus comprehensive documentation, but I've found a ton of value in [dbt](https://www.getdbt.com/), an open source tool that does all this for you. They recently added beta support for python so you wouldn't even need to change your code base all that much. In short, it would be a free and open source tool to help you:
- structure all your files
- add unit, integration, and functional tests
- create documentation
- automatically create lineage graphs between scripts
- and other things I'm not thinking of right now

I promise I'm not paid by dbt, but it has made my life a lot easier while building out data warehouses 😄. [Dagster also has an article](https://dagster.io/blog/dagster-dbt) on how it integrates with dbt.

Let me know if I missed anything, I'm more than happy to discuss this further!

**Reply by tmylk** [reply]
(Apologies for the unsolicited 2 cents.)

I have had great experience building pipelines in dbt. And I like the UI and hosted-ness of DBT cloud a lot.

Main advantage of DBT is that it is SQL, and it is bite-sized small steps, so it widens the number of people who can contribute to maintaining the pipeline. Now anyone with basic SQL can chip in, not just a Python data engineer.

 I imagine some of your users are technical enough to grok some SQL. Wouldn't it be nice if a domain expert in gov stats can fix the renames, change units etc themselves independently. That frees data engineer time for more complex tasks.

**Comment by TrentonBush** [comment]
I like the promise of Dagster to solve all of the "visibility into intermediate outputs" issues via caching. *But* it also forces people to access PUDL via the development python package and pandas, which I speculate is a high bar that would drastically cut down the number of users. Based on that intuition, I'd lean towards a more limited - but easier to distribute and use - database-based caching method (AKA data warehouse/marts). It'll take some careful consideration to pick the right points in the DAG, but it seems tractable.

**Reply by bendnorman** [reply]
With Dagster, we can [Materialize Assets](https://docs.dagster.io/concepts/assets/asset-materializations) to keep track of where data assets are created in the DAG. The assets can be written to a database so the intermediate tables can be easily distributed. I don't think Asset Materializations force people to access PUDL via the python package. Asset Materializations allow users to see where a given table is created in the DAG.

**Reply by zaneselvans** [reply]
@TrentonBush the idea here is that all of the useful outputs that are currently either transient (like the raw extracted tables) or require the Python package (like the MCOE, state level demand allocations, denormalized output tables) would be written to the database, and thus available without Python at all, or via the minimal dependencies that the Intake catalogs require.

**Reply by TrentonBush** [reply]
Oh ok that's way better than I thought. Long live Dagster!

**Reply by MichaelTiemannOSC** [reply]
The Data Commons architecture project of the OS-Climate team (https://github.com/os-climate/os_c_data_commons/blob/main/os-c-data-commons-architecture-blueprint.md) has been working on this from the opposite end (being a platform to federate/maintain/distribute data), and have made some progress with dagster, dbt, open metadata, etc.  I cannot promise anything yet, but I will circulate your thoughts with our team for discussion about potential collaboration.  Your ETL work is aces, and there's a lot of potential in the federated governance model we are building (which needs data to demonstrate).
